org.jacoco.core/src/org/jacoco/core/internal/instr/ProbeInserter.java

```diff
@@ -68,27 +67,69 @@
 		variable = pos;
 	}
 
+	/**
+	 * Insert bytecode (a probe) to increment the position corresponding to the
+	 * given {@code id} in the int[] array
+	 *
+	 * @param id
+	 *            the position to increment
+	 */
 	public void insertProbe(final int id) {
-
-		// For a probe we set the corresponding position in the boolean[] array
-		// to true.
-
+		// Retrieve the int[] containing coverage information
 		mv.visitVarInsn(Opcodes.ALOAD, variable);
 
-		// Stack[0]: [Z
-
+		// Stack[0]: [I
+		// Pushes the index of the array we want to retrieve on the stack
 		InstrSupport.push(mv, id);
 
 		// Stack[1]: I
-		// Stack[0]: [Z
+		// Stack[0]: [I
+		// Duplicate the stack as we want to read and write to the array
+		// with the index.
+		mv.visitInsn(Opcodes.DUP2);
+
+		// Stack[3]: I
+		// Stack[2]: [I
+		// Stack[1]: I
+		// Stack[0]: [I
+		// Retrieve the value from the array
+		mv.visitInsn(Opcodes.IALOAD);
 
+		// Stack[2]: I
+		// Stack[1]: I
+		// Stack[0]: [I
+		// Push 1 to the stack
 		mv.visitInsn(Opcodes.ICONST_1);
 
+		// Stack[3]: I
+		// Stack[2]: I
+		// Stack[1]: I
+		// Stack[0]: [I
+		// Increment the integer on the stack
+		mv.visitInsn(Opcodes.IADD);
+
 		// Stack[2]: I
 		// Stack[1]: I
-		// Stack[0]: [Z
+		// Stack[0]: [I
+		// Load the max integer value - 1 on the stack. We don't use max integer
+		// as Math.min with incremented max int would be negative.
+		mv.visitLdcInsn(0x7ffffffe);
 
-		mv.visitInsn(Opcodes.BASTORE);
```

Comment:
Just curious: how does Jacoco handle thread safety? It's not clear that the old code was thread-safe (IIUC it wasn't, either), but the new code depends more on the original value in the array than the old code.

Reply:
Hmm good comment. [According to their docs](https://www.jacoco.org/jacoco/trunk/doc/flow.html) (see section Probe Implementation) the probe implementation is to be thread-safe and they mention explicitly that it is in that section. 

I guess that this new code is indeed not thread-safe due to the way that we indeed depend on the value in the array. I'm not sure if it is easy to fix that without introducing more overhead. In any case, I don't expect major problems from this apart from missing hits when two threads retrieve the value from the array before one could store the incremented value. I believe the cap-implementation will still work to prevent the integer from overflowing even when multiple threads run this bytecode. Therefore I believe we can accept this where I will need to report it as a threat to validity in my thesis. Do you perhaps have some other thoughts on this?

Reply:
As discussed in private, the code is indeed not thread safe. A synthetic test mimicking implementation behavior (both incrementing with overflow, capping using `Math.min` and capping using double casting) has shown that if 12 threads (= virtual cores) increment a value at the same index 1000 times that about 65K-75K out of 120K increments are missed. If we increment values at either index 0 or 1 (randomly) using the same test set-up, the amount of misses has decreased to 50K-55K out of 120K. Randomly to index 0-5 results in 20K-35K, 0-10 results in 15K-20K, 0-100 in 1K-2K.

The lack of thread safety has only resulted in missing these hits, no other side effects. We have accepted that for our project, we accept this threat to validity as we suspect it will not impact our analysis too much. This synthetic scenario is not likely to happen due to the following points:
- Most often, available threads execute different code paths as part of different incoming HTTP requests.
- Even if multiple running threads execute the same code path, they can differ in start times such that this conflict will not occur. 
- The chance that these conflicts happen is higher with code that is executed more often. As our analysis is focused on comparing code that is executed more often than code that is executed less often (say once versus 100K times), a few misses won't impact that difference too much.